server/neptune/workflows/internal/deploy/workflow.go

```diff
@@ -0,0 +1,168 @@
+	RevisionReceiver         SignalReceiver
+	NewRevisionSignalChannel workflow.ReceiveChannel
+}
+
+func newRunner(ctx workflow.Context, request Request) *Runner {
+	// convert to internal types, we should probably move these into another struct
+	repo := github.Repo{
+		Name:     request.Repository.Name,
+		Owner:    request.Repository.Owner,
+		URL:      request.Repository.URL,
+	}
+
+	// inject dependencies
+
+	// temporal effectively "injects" this, it just cares about the method names,
+	// so we're modeling our own DI around this.
+	var a *activities.Deploy
+
+	revisionQueue := queue.NewQueue()
+	revisionReceiver := revision.NewReceiver(ctx, revisionQueue)
+
+	worker := &queue.Worker{
+		Queue:      revisionQueue,
+		Activities: a,
+		Repo:       repo,
+		RootName:   request.Root.Name,
+	}
+
+	return &Runner{
+		QueueWorker:              worker,
+		RevisionReceiver:         revisionReceiver,
+		NewRevisionSignalChannel: workflow.GetSignalChannel(ctx, NewRevisionSignalID),
+	}
+}
+
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action RunnerAction
+	workerCtx, shutdownWorker := workflow.WithCancel(ctx)
+
+	wg := workflow.NewWaitGroup(ctx)
+	wg.Add(1)
+
+	// if this panics in anyway, we'll need to ship a fix to the running workflows, else risk dropping
+	// signals
+	// should we have some way of persisting our queue in case of workflow termination?
+	// Let's address this in a followup
+	workflow.Go(workerCtx, func(ctx workflow.Context) {
+		defer wg.Done()
+		r.QueueWorker.Work(ctx)
+	})
+
+	onTimeout := func(f workflow.Future) {
+		err := f.Get(ctx, nil)
+
+		if temporal.IsCanceledError(err) {
+			action = OnCancel
+			return
+		}
+
+		action = OnTimeout
+	}
+
+	s := temporalInternal.SelectorWithTimeout{
+		Selector: workflow.NewSelector(ctx),
+	}
+	s.AddReceive(r.NewRevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		r.RevisionReceiver.Receive(c, more)
+		action = OnReceive
+
+	})
+	cancelTimer, _ := s.AddTimeout(ctx, RevisionReceiveTimeout, onTimeout)
+
+	// main loop which handles external signals
+	// and in turn signals the queue worker
+	for {
+		// blocks until a configured callback fires
+		s.Select(ctx)
+
+		switch action {
+		case OnCancel:
+			continue
+		case OnReceive:
+			cancelTimer()
+			cancelTimer, _ = s.AddTimeout(ctx, RevisionReceiveTimeout, onTimeout)
+			continue
+		}
+
+		logger.Info(ctx, "revision receiver timeout")
+
+		// check state here since if we timed out, we're probably not susceptible to the queue
+		// worker being in a waiting state right before it's about to start working on an item.
+		if !s.HasPending() && r.QueueWorker.GetState() != queue.WorkingWorkerState {
+			shutdownWorker()
+			break
+		}
+
+		// basically keep on adding timeouts until we can either break this loop or get another signal
+		// we need to use the timeoutCtx to ensure that this gets cancelled when when the receive is ready
+		cancelTimer, _ = s.AddTimeout(ctx, RevisionReceiveTimeout, onTimeout)
```

Comment:
What are all the timeouts for? Is the intent to let the worker exit between signals? Or just to keep staying alive but periodically check to see if we're cancelled? If it's the latter, wouldn't it be enough to select (again) between just the revision receiver and cancellation on the workflow context (no timeout involved)?

Reply:
Separately, should we be cancelling the existing timer before overwriting it with a new one?